csds-client/client/util.go

```diff
@@ -0,0 +1,201 @@
+package client
+
+import (
+	"encoding/json"
+	"fmt"
+	envoy_api_v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	envoy_config_filter_http_router_v2 "github.com/envoyproxy/go-control-plane/envoy/config/filter/http/router/v2"
+	envoy_config_filter_network_http_connection_manager_v2 "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"
+	envoy_service_status_v2 "github.com/envoyproxy/go-control-plane/envoy/service/status/v2"
+	envoy_type_matcher "github.com/envoyproxy/go-control-plane/envoy/type/matcher"
+	"github.com/ghodss/yaml"
+	"github.com/golang/protobuf/proto"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+// helper method for parsing csds request yaml to nodematchers
+func ParseYaml(path string, yamlStr string, nms *[]*envoy_type_matcher.NodeMatcher) error {
+	if path != "" {
+		// parse yaml to json
+		filename, _ := filepath.Abs(path)
+		yamlFile, err := ioutil.ReadFile(filename)
+		if err != nil {
+			return fmt.Errorf("%v", err)
+		}
+		js, err := yaml.YAMLToJSON(yamlFile)
+		if err != nil {
+			return fmt.Errorf("%v", err)
+		}
+
+		// parse the json array to a map to iterate it
+		var data map[string]interface{}
+		err = json.Unmarshal(js, &data)
+		if err != nil {
+			return fmt.Errorf("%v", err)
+		}
+
+		// parse each json object to proto
+		for _, n := range data["node_matchers"].([]interface{}) {
+			x := &envoy_type_matcher.NodeMatcher{}
+
+			jsonString, err := json.Marshal(n)
+			if err != nil {
+				return fmt.Errorf("%v", err)
+			}
+			err = protojson.Unmarshal(jsonString, x)
+			if err != nil {
+				return fmt.Errorf("%v", err)
+			}
+			*nms = append(*nms, x)
+		}
+	}
+	if yamlStr != "" {
+		var js []byte
+		var err error
+		// json input
+		if yamlStr[0] == '{' {
```

Comment:
I might be ignorant here, but just to confirm - is this the reliable/standard method to auto-detect json? Maybe there are some libraries exist we could use instead?

Reply:
Added a function `isJson()` to call `json.NewDecoder()` to validate json format string.